### PR TITLE
Rephrase E2E, Performance, and Visual test sections

### PR DESCRIPTION
**Pull request by bubafinder**
Rephrase E2E, Performance, and Visual test sections with minor changes and mostly extend the `Antipatterns` and `Best practices` sections.

### CODE REVIEW

**Comment by bubafinder** [comment]
@droguljic can you please add someone from the QA team as a Reviewer? I don't see anyone on the list, and as I can see, you have permission to add them.

File: recipes/automated-testing.md [hunk 1]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
-run. If we need to test edge cases, we should try to implement those at a lower
-level (integration or unit tests).
+These tests should not cover all of the use cases because they are the slowest
+to run. If we need to test edge cases, we should try to implement those at a
+lower level (integration or unit tests).
 
 #### When to use
-- Test user interaction with the application UI.
+- To validate user interactions and critical workflows in the application UI.
+- For testing full system integration across multiple services or components.
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:194]: @bubafinder this can be an antipattern - for example in a microservice environment, or when external services are used.
For once, the effect of integration through network communication significantly impacts flakiness. Instead, combining other types of tests, like contract-based ones, would be a better approach.
Alternatively, different components can be mocked, but this poses other challenges.

Due to this, I'd remove this line.

File: recipes/automated-testing.md [hunk 2]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
 
 #### When **not** to use
 - For data validation.
 
 #### Best practices
+- Focus on the most important user workflows rather than attempting exhaustive
+coverage.
+- Each test should be able to run independently, with the environment reset to a
+known state before every test.
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:203]: Great points 👍

File: recipes/automated-testing.md [hunk 3]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
-
+- Ensure the test data mirrors real-world conditions, including varying user
+inputs and dataset sizes.
 
 ### Visual Tests
 
-The type of test where test runner navigates to browser page, takes screenshot
-and then compares the future screenshots with the reference screenshot.
+The type of test where test runner navigates to browser page, takes snapshot and
+then compares the snapshots with the reference snapshot.
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:272]:
```suggestion
The type of test where test runner navigates to browser page, takes snapshot and
then compares the snapshot with the reference snapshot.
```

File: recipes/automated-testing.md [hunk 4]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
+Visual tests allow you to quickly cover large portions of the application,
+ensuring that changes in the UI are detected without writing complex test cases.
+The downside is that they're requiring engineers to invest time in identifying
+the root cause of errors.
 
 #### When to use
-- When we want to cover broad range of features.
-- When we want to increase test coverage with least effort.
 - When we want to make sure there are no changes in the UI.
+- When we want to increase test coverage with least effort.
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:281]: @mbareta going through this again (this is not edited in this PR), I'm thinking whether we should avoid suggesting this and the next point altogether to avoid motivating people to take the shortcut in testing. 
What do you think?

Comment (MiroDojkic) [reply recipes/automated-testing.md:281]: It's not even less of an effort given almost every change in the early stage of development needs to be manually approved. 😅

File: recipes/automated-testing.md [hunk 5]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
+- During the early stages of a project, while E2E tests are still under
+development.
 
 #### When **not** to use
 - To test a specific feature or business logic.
 - To test a specific user scenario.
 
 #### Best practices
-- Have deterministic seeds so the UI always renders the same output.
+- Ensure the UI consistently renders the same output by controlling randomness
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:290]: This line is more abstract and ambiguous to me than the previous one. Would it make sense to say:
`Ensure the UI consistently renders the same output by eliminating randomness
(e.g., by always using same seeds data or controlling API responses to always return same values).`

File: recipes/automated-testing.md [hunk 6]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
 - Add as many pages as possible but keep the tests simple.
+- Consider running visual tests at the component level to isolate and detect
+issues earlier.
+- Define acceptable thresholds for minor visual differences (e.g., pixel
+tolerance) to reduce noise while detecting significant regressions.
 
 #### Antipatterns
+- Avoid creating overly complicated visual tests that try to simulate user
+behavior. These are better suited for E2E testing.
+- Visual tests should complement, not replace other types of tests like E2E
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:301]: 👌

File: recipes/automated-testing.md [hunk 7]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
+backend systems or environments.
+- Run tests in parallel and isolate them from external dependencies to improve
+speed and reliability.
+- Automate E2E tests in your CI/CD pipeline to catch regressions early in the
+deployment process.
 
 #### Antipatterns
+- Avoid trying to cover all use cases or edge cases in E2E tests; these are
+better suited for unit or integration tests.
+- Don’t bypass or disable flaky tests without fixing them, as they undermine the
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:219]: Since this section is antipatterns, it'd make more sense to phrase points to say what an antipattern is, i.e.
```suggestion
- Bypassing or disabling flaky tests only due to flakiness - this reduces coverage and confidence
```

Or to offer alternative option - does it make sense to even have antipatterns if we already have the same things in best practices?
I feel like antipatterns would be a better fit in examples. @mbareta  @bubafinder

File: recipes/automated-testing.md [hunk 8]
```diff
@@ -174,91 +174,131 @@ performance tests (K6).
...
 and good performance will allow that.
 - Flaky tests should be immediately disabled and refactored. Flaky tests will
-cause the team to ignore or bypass the tests and these should be dealt with immediately.
+cause the team to ignore or bypass the tests and these should be dealt with
+immediately.
+- Ensure consistent data states to avoid test failures due to variability in
+backend systems or environments.
+- Run tests in parallel and isolate them from external dependencies to improve
+speed and reliability.
+- Automate E2E tests in your CI/CD pipeline to catch regressions early in the
```
Comment (MiroDojkic) [inline recipes/automated-testing.md:213]: Any ideas on what to suggest to devs when to execute E2E tests?
Every PR/every commit/every release?

File: recipes/automated-testing.md (hunk 5 above)
Comment (bubafinder) [reply recipes/automated-testing.md:290]: You are right

File: recipes/automated-testing.md (hunk 7 above)
Comment (bubafinder) [reply recipes/automated-testing.md:219]: I agree, it's mostly covered with one point in the best practices section. I will remove it. It would be hard to cover flaky tests with an example 🤔

File: recipes/automated-testing.md (hunk 8 above)
Comment (bubafinder) [reply recipes/automated-testing.md:213]: We already discussed this, and I don't think we have a good recipe for that. It depends on the project, the architecture, the process, as well as the team itself. I would rather leave it up to the team to choose.

Comment (bubafinder) [reply recipes/automated-testing.md:213]: Maybe we can make a few examples once we start to work on the examples section, with explained advantages and disadvantages for each individual.

File: recipes/automated-testing.md (hunk 4 above)
Comment (bubafinder) [reply recipes/automated-testing.md:281]: I'm aware that this may be motivation for someone to do something the easy way, but I think we have enough capable teams to make a reasonable decision. This is a fact and It is up to developers/teams to decide how they will use it.

File: recipes/automated-testing.md (hunk 1 above)
Comment (bubafinder) [reply recipes/automated-testing.md:194]: I would rather keep this because there are mechanisms where flakiness caused by these things can be controlled, but the benefit of such tests is huge in some cases. This implies that the tests should be atomic and simple, all complicated tests should be thrown out. Likewise, if such tests cover crucial features of the project, they are certainly important (for example payment process on the Solution Tree project). I think that this is also something where the team needs to have a good feeling to know what and how to cover.

File: recipes/automated-testing.md [hunk 9]
```diff
@@ -174,91 +174,133 @@ performance tests (K6).
...
+points, and track improvements.
+- Performance testing should not be a one-time task; it should be an ongoing
+process integrated into the development lifecycle.
 
 #### Antipatterns
 - Running these tests locally or on an environment that doesn't match production
 in terms of infrastructure performance. (tests should be developed on a local
 instance, but the actual measurements should be performed live)
-
+- Ensure the test data mirrors real-world conditions, including varying user
```
Comment (mbareta) [inline recipes/automated-testing.md:267]: Why is this an antipattern?

File: recipes/automated-testing.md (hunk 4 above)
Comment (mbareta) [reply recipes/automated-testing.md:282]: @MiroDojkic I agree. It sounds like shitty marketing. Maybe we can rephrase it so is sounds like:
> When we want to cast a wide net and cover lots of code with small amount of tests.
But then we should emphasize that:
> Due to wide scope of these tests, they have very low __determinism__ and they shouldn't be used to test core functionality of the app (for example: logging in, payments, algorithms, etc.).
What do you say?

File: recipes/automated-testing.md (hunk 1 above)
Comment (mbareta) [reply recipes/automated-testing.md:194]: @bubafinder I'd remove this line in its current form. For general testing guidelines, I'd say `full system integration` belongs in integration tests. Although I understand where this is coming from - you meant that we can test the entire stack on a specific use-case. I think we got this covered by writing this part https://github.com/ExtensionEngine/tech-guide/pull/13/files#diff-3594683eea2e93dba4848677618559175ae68355dacd365b56050cd1a161cf27R180
Solution Tree's payment process is a specific use case tested in E2E tests by clicking on the UI.

Instead we can add:
> For testing specific user flows.
> For making sure that critical application features are working as expected.
> For better coverage of the most common user pathways.

And also in the Best Practices section:
> Tests should be atomic and simple, all complicated tests should be thrown out.

I know this is repetitive and maybe we don't need it, but I feel like this covers what you wanted to say better than this line which can be interpreted in different ways.

File: recipes/automated-testing.md (hunk 4 above)
Comment (MiroDojkic) [reply recipes/automated-testing.md:282]: > It sounds like shitty marketing.
That's a harsh thing to say. 😅
I don't see it being bad, but before I answer your question, I'd like to zoom out and ask both of you - what behavior are we trying to accomplish from devs and QAs by mentioning this?

Comment (mbareta) [reply recipes/automated-testing.md:282]: We talked about tests in general and I heard a few times that visual tests are the easiest way to introduce tests to existing application which doesn't have tests. I don't want to guide people or suggest we should start with visual testing first so I'd rephrase it. But my intent was to highlight that visual tests are the easiest way to bring meaningful test coverage to an existing large app.

File: recipes/automated-testing.md [hunk 10]
```diff
@@ -174,91 +174,133 @@ performance tests (K6).
...
+inputs and dataset sizes.
 
 ### Visual Tests
 
-The type of test where test runner navigates to browser page, takes screenshot
-and then compares the future screenshots with the reference screenshot.
+The type of test where test runner navigates to browser page, takes snapshot and
```
[the remark this answers is not in the document]
Comment (bubafinder) [reply recipes/automated-testing.md:272]: I am using snapshot word intentionally because most tools used for visual testing are actually taking a snapshot (not a screenshot) of the code, together with CSS and JS, which is then interpreted in the browsers on their side.

File: recipes/automated-testing.md (hunk 4 above)
Comment (bubafinder) [reply recipes/automated-testing.md:282]: Same thing on my side, to highlight that visual tests are the easiest way to bring meaningful test coverage to an existing large app. What is a good thing if it is used properly.

File: recipes/automated-testing.md (hunk 9 above)
Comment (bubafinder) [reply recipes/automated-testing.md:267]: I was moving it up and down between best practices and antipatterns but didn't rephrase it at the end.

**Review by mbareta — APPROVED** [review]
